### PR TITLE
Update trade interface to better inform players of consequences

### PR DESCRIPTION
**Pull request by pjf**
#### Summary

```SUMMARY: Interface "NPC trades will check for confirmation and inform players if they may not get full value."```

#### Purpose of change

- It was possible for the player to accidentally commit to a trade just by pressing 'enter', which is also used when typing item quantities.
- NPCs would limit how much they were willing to owe the player, but players were never informed of this.
- NPCs would be willing to give the player some credit, but players were never informed of this, either.
- Existing code was not always clear in workings or intent.

#### Describe the solution

Code changes which:

- Asks the player to confirm any trade.
- Tells the player know if they're going to not get full credit for the trade.
- Tells the player if they could go into a little debt, but are asking too much.

Internally, we also have:

- NPC method to check if NPCs `will_exchange_items_freely`.
- NPC methods for `max_credit_extended` and `max_willing_to_owe`.
- Renaming `p` to `np` in multiple places for clarity.
- Renamed `u_get` to `your_balance`. You get no money from these trades atm, but NPCs remember debts and credit.
- Helper variables for color formatting for improved clarity.
- An `npc_will_accept_trade` method to make it clear under what conditions a trade can occur.
- A `calc_npc_owes_you` method to determine how much an NPC would owe you if the trade went ahead.
- General tidying up.

#### Describe alternatives you've considered

Just playing the game instead of stopping to fix any pain points I encounter. :)

#### Additional context

<s>This PR is based on top of #33132 and thus incorporates its commits. Merging this PR will also merge that PR automatically.</s>

This PR has been rebased onto master as #33132 has been merged.

#### Screenshots

trade1

trade2

### CODE REVIEW

**Comment by RarkGrames** [comment]
Flashbacks from Fallout when trading 25 FN FALs for a stimpack... :laughing:

**Comment by Maddremor** [comment]
Why would NPCs not be willing to owe the player beyond a certain sum? If they were willing to buy the item, deferring payment to a later date would not cost them anything. If the player died, they would have gotten the item for free.

**Comment by pjf** [comment]
> Why would NPCs not be willing to owe the player beyond a certain sum?

This is me maintaining existing behaviour. NPCs already cap the amount they're willing to owe you, but they just don't tell you, making it easy for the player to waste value in a trade by accident.

The calculations for how much an NPC is willing to owe or be owed are now in methods on the `npc` class so it's easy to change in the future.

**Comment by pjf** [comment]
jenkins rebuild

File: src/npc.h [hunk 1]
```diff
@@ -877,6 +877,11 @@ class npc : public player
         bool wants_to_buy( const item &it ) const;
         bool wants_to_buy( const item &/*it*/, int at_price, int /*market_price*/ ) const;
 
+        // Will swap items with player at no charge. (Eg: NPC ally)
```
Comment (BevapDin) [inline src/npc.h]: Please don't put documentation for functions in the header **and** in the cpp file. This is very confusing. A reader will typically expected the documentation at either place, but not at both. A future coder may also easily forget to update the documentation at *both* places.

File: src/npctrade.cpp [hunk 2]
```diff
@@ -442,20 +444,46 @@ bool trading_window::perform_trade( npc &p, const std::string &deal )
...
                 } else if( volume_left < 0_ml || weight_left < 0_gram ) {
                     // Make sure NPC doesn't go over allowed volume
-                    popup( _( "%s can't carry all that." ), p.name );
+                    popup( _( "%s can't carry all that." ), np.name );
                     update = true;
                     ch = ' ';
+                } else if( calc_npc_owes_you( np ) < your_balance ) {
+                    // NPC is happy with the trade, but isn't willing to remember the whole debt.
+                    const bool trade_ok = query_yn(
+                                              string_format(
```
Comment (BevapDin) [inline src/npctrade.cpp]: `query_yn` does string formatting, it's part of its job. Please don't steal its job.

File: src/npctrade.cpp [hunk 3]
```diff
@@ -551,11 +601,17 @@ bool npc_trading::trade( npc &np, int cost, const std::string &deal )
         }
 
         // NPCs will remember debts, to the limit that they'll extend credit or previous debts
-        if( !trade_win.exchange ) {
+        if( ! np.will_exchange_items_freely() ) {
```
Comment (BevapDin) [inline src/npctrade.cpp]: Btw. the space after the `!` is very unusual. It's not used in the existing code base.

File: src/npctrade.cpp (hunk 2 above)
Comment (pjf) [reply src/npctrade.cpp]: Oh! I must have missed that. This makes me very happy!

File: src/npctrade.cpp (hunk 3 above)
Comment (pjf) [reply src/npctrade.cpp]: I agree that no-space is more common in the CDDA codebase, but a quick grep shows a number of files which have the space-included version, and I think the difference in readability is slight.

**Comment by pjf** [comment]
Code updated based upon code review. Thank you, @BevapDin !

**Comment by pjf** [comment]
jenkins rebuild

**Comment by pjf** [comment]
jenkins rebuild

**Pull request merged**